File: serve.js
// Icy's Second Brain - local server (no dependencies).
// Run:   node serve.js      then open  http://localhost:4321
// It serves the site/ pages and saves captured notes straight into your vault.
// Change the vault location with:  OBSIDIAN_VAULT_PATH=/path/to/icy-vault node serve.js
const http = require('http');
const fs = require('fs');
const path = require('path');
const os = require('os');

const isPkg = !!process.pkg;                                  // true when running as the packaged .exe
const BASE = isPkg ? path.dirname(process.execPath) : __dirname; // folder the exe/script lives in
const SITE = path.join(BASE, 'site');
const VAULT = process.env.OBSIDIAN_VAULT_PATH || path.join(BASE, 'icy-vault');
const PORT = process.env.PORT || 4321;
function lanIP(){ const ifs=os.networkInterfaces(); for(const n of Object.keys(ifs)){ for(const i of ifs[n]){ if(i.family==='IPv4'&&!i.internal) return i.address; } } return null; }

const MIME = { '.html': 'text/html; charset=utf-8', '.js': 'text/javascript', '.css': 'text/css',
  '.json': 'application/json', '.png': 'image/png', '.svg': 'image/svg+xml', '.ico': 'image/x-icon',
  '.jpg': 'image/jpeg', '.webp': 'image/webp' };

function send(res, code, body, type) {
  res.writeHead(code, { 'Content-Type': type || 'text/plain; charset=utf-8' });
  res.end(body);
}

const server = http.createServer((req, res) => {
  const u = new URL(req.url, 'http://localhost');

  // Save a note into the vault
  if (req.method === 'POST' && u.pathname === '/api/save') {
    let data = '';
    req.on('data', c => { data += c; if (data.length > 2e6) req.destroy(); });
    req.on('end', () => {
      let body = {};
      try { body = JSON.parse(data); } catch (e) {}
      const rel = String(body.path || '').replace(/\.\.+/g, '').replace(/^[\/\\]+/, '');
      if (!rel || !body.content) return send(res, 400, JSON.stringify({ error: 'path and content required' }), 'application/json');
      const dest = path.resolve(VAULT, rel);
      if (!dest.startsWith(path.resolve(VAULT))) return send(res, 400, JSON.stringify({ error: 'bad path' }), 'application/json');
      fs.mkdir(path.dirname(dest), { recursive: true }, err => {
        if (err) return send(res, 500, JSON.stringify({ error: String(err) }), 'application/json');
        fs.writeFile(dest, body.content, e => {
          if (e) return send(res, 500, JSON.stringify({ error: String(e) }), 'application/json');
          console.log('  saved ->', dest);
          send(res, 200, JSON.stringify({ ok: true, url: 'file:///' + dest.replace(/\\/g, '/') }), 'application/json');
        });
      });
    });
    return;
  }

  // Static files from site/
  let p = u.pathname === '/' ? '/index.html' : u.pathname;
  let file = path.resolve(SITE, '.' + decodeURIComponent(p));
  if (!file.startsWith(path.resolve(SITE))) return send(res, 403, 'forbidden');
  fs.readFile(file, (err, buf) => {
    if (err) return send(res, 404, 'not found');
    send(res, 200, buf, MIME[path.extname(file)] || 'application/octet-stream');
  });
});

server.listen(PORT, () => {
  console.log('');
  console.log('  Icy Second Brain is running:  http://localhost:' + PORT);
  const ip = lanIP(); if (ip) console.log('  On your phone (same WiFi):    http://' + ip + ':' + PORT);
  console.log('  Site:   ' + SITE);
  console.log('  Vault:  ' + VAULT + '   (notes save here)');
  console.log('  Stop by closing this window (or Ctrl+C)');
  console.log('');
  const _url = 'http://localhost:' + PORT;
  const _open = process.platform === 'win32' ? 'start ""' : process.platform === 'darwin' ? 'open' : 'xdg-open';
  try { require('child_process').exec(_open + ' ' + _url); } catch (e) {}
});

File: serve.py
#!/usr/bin/env python3
# Icy's Second Brain - local server (Python, no dependencies).
# Run:   python serve.py     then open  http://localhost:4321
# Saves captured notes straight into your vault (icy-vault next to this file,
# or set OBSIDIAN_VAULT_PATH).
import os, json, socket, socketserver, http.server

BASE = os.path.dirname(os.path.abspath(__file__))
SITE = os.path.join(BASE, 'site')
VAULT = os.environ.get('OBSIDIAN_VAULT_PATH', os.path.join(BASE, 'icy-vault'))
PORT = int(os.environ.get('PORT', '4321'))


class Handler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *a, **k):
        super().__init__(*a, directory=SITE, **k)

    def _json(self, code, obj):
        b = json.dumps(obj).encode('utf-8')
        self.send_response(code)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(b)))
        self.end_headers()
        self.wfile.write(b)

    def do_POST(self):
        if self.path.split('?')[0] != '/api/save':
            return self.send_error(404)
        n = int(self.headers.get('Content-Length', 0) or 0)
        try:
            body = json.loads(self.rfile.read(n) or b'{}')
        except Exception:
            body = {}
        rel = str(body.get('path', '')).replace('..', '').lstrip('/\\')
        content = body.get('content')
        if not rel or not content:
            return self._json(400, {'error': 'path and content required'})
        dest = os.path.normpath(os.path.join(VAULT, rel))
        if not dest.startswith(os.path.normpath(VAULT)):
            return self._json(400, {'error': 'bad path'})
        os.makedirs(os.path.dirname(dest), exist_ok=True)
        with open(dest, 'w', encoding='utf-8') as f:
            f.write(content)
        print('  saved ->', dest)
        return self._json(200, {'ok': True, 'url': 'file:///' + dest.replace('\\', '/')})

    def log_message(self, *a):
        pass


def lan_ip():
    try:
        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        s.connect(('8.8.8.8', 80))
        ip = s.getsockname()[0]
        s.close()
        return ip
    except Exception:
        return None


if __name__ == '__main__':
    socketserver.ThreadingTCPServer.allow_reuse_address = True
    httpd = socketserver.ThreadingTCPServer(('0.0.0.0', PORT), Handler)
    print('')
    print('  Icy Second Brain is running:  http://localhost:%d' % PORT)
    ip = lan_ip()
    if ip:
        print('  On your phone (same WiFi):    http://%s:%d' % (ip, PORT))
    print('  Vault:  %s   (notes save here)' % VAULT)
    print('  Stop with Ctrl+C')
    print('')
    try:
        import webbrowser
        webbrowser.open('http://localhost:%d' % PORT)
    except Exception:
        pass
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        pass
